netconf_client/ncclient.py

```diff
@@ -21,40 +24,258 @@
-    def __init__(self, session, timeout=120):
+    def __init__(self, session, timeout=120, log_id=None):
         """Construct a new Manager object
 
         :param session: The low-level NETCONF session to use for requests
         :type session: :class:`netconf_client.session.Session`
 
         :param float timeout: Duration in seconds to wait for replies
+        :param string log_id: log ID string printed with each log entry,
+               the peer's IP address as default
         """
         self.timeout = timeout
         self.session = session
+        self.log_id = log_id
+        self.start_time = self._get_timestamp()
+        self.log_level = logging.DEBUG
+
+        self.logged_functions = {
+            "get": True,
+            "edit_config": True,
+            "get_config": True,
+            "copy_config": True,
+            "discard_changes": True,
+            "commit": True,
+            "lock": True,
+            "unlock": True,
+            "kill_session": True,
+            "create_subscription": True,
+            "validate": True,
+            "delete_config": True,
+            "dispatch": True,
+            "close_session": False,
+        }
 
     def __enter__(self):
         return self
 
     def __exit__(self, a, b, c):
         self.session.__exit__(a, b, c)
 
+    @staticmethod
+    def get_logger():
+        """Returns the internally used logger instance (same for all sessions)"""
+        return logger
+
+    def get_log_level(self):
+        """Returns current logging level, default is DEBUG"""
+        return self.log_level
+
+    def set_log_level(self, level=logging.DEBUG):
+        """Sets the log level to use for logging"""
+        self.log_level = level
+
+    def get_logged_functions(self):
+        """Returns a dict 'str'->'Boolean' of function names to be logged"""
+        return self.logged_functions
+
+    def is_function_logged(self, funcname):
+        """Checks whether a given function is enabled for logging"""
+        return self.logged_functions.get(funcname, False)
+
+    def enable_logging(self, funcname):
+        """Enables logging for given function name"""
+        if funcname in self.logged_functions:
+            self.logged_functions.update({funcname: True})
+
+    def disable_logging(self, funcname):
+        """Disables logging for given function name"""
+        if funcname in self.logged_functions:
+            self.logged_functions.update({funcname: False})
+
+    def _get_timestamp(self):
+        return datetime.now()
+
+    def _is_logger_enabled(self, funcname):
+        return self.is_function_logged(funcname) and Manager.get_logger().isEnabledFor(
+            self.get_log_level()
+        )
+
+    def _get_peer_ip(self):
+        """Returns the peer's IP address.
+           If not connected, an empty string is returned.
+        """
+
+        addr = ""
+        try:
+            s = self.session.sock.sock
+            (addr, _) = s.getpeername()
+        except:
```

Comment:
Bare `except` is generally dangerous. This should catch the specific errors that may be raised.

Reply:
Yes, you are right. However, it does no matter what exceptions might be thrown, because the reaction is always the same: "no peer info avaiable". So I could try to list and catch all possible exception, but it would not provide any additional benefit.

Reply:
The reasoning for this is e.g. there could be a typo in the line above and we wouldn't know. Is `self.session.sock.sock` the correct attribute? What if we change it without changing this code? The `AttributeError` would be missed and this function wouldn't be doing what we actually need it to do.

This is a well known [anti-pattern][1] in Python. There are a few rare cases where it is appropriate (e.g. a completely generic "retry" function). But these cases are very rare.

[1]: https://mike.pirnat.com/2009/05/09/the-diaper-pattern-stinks/

Reply:
I have already forgotten that Python is an interpreter and not a C/C++ compiler ;-). I'll look for the set of possible exceptions and add the according `except` handlers.